### PR TITLE
Add Substr built-in function

### PR DESCRIPTION
**Pull request by luochenUmich**
This PR add support for "substr" built-in function in codegen engine. The log shows that we correctly calculate string length and starting pointer. Some test cases are not working because of the varchar length bug.

### CODE REVIEW

**Comment by hzxa21** [comment]
This PR should be merged after #911

**Comment by pmenon** [comment]
@luochenUmich Can you rebase this?

**Comment by apavlo** [comment]
@luochenUmich Hey I am sorry about your bowel problems, but do you think that you can rebase this?

**Comment by luochenUmich** [comment]
@apavlo Just rebased

File: src/codegen/type/varchar_type.cpp [hunk 1]
```diff
@@ -11,7 +11,6 @@
 //===----------------------------------------------------------------------===//
 
 #include "codegen/type/varchar_type.h"
-
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Undo

File: src/codegen/type/varchar_type.cpp [hunk 2]
```diff
@@ -212,6 +211,7 @@ struct DateTrunc : public TypeSystem::BinaryOperatorHandleNull {
   }
 };
 
+
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Remove empty line

File: src/codegen/type/varchar_type.cpp [hunk 3]
```diff
@@ -303,6 +303,32 @@ struct RTrim : public TypeSystem::BinaryOperatorHandleNull {
   }
 };
 
+struct Substr : public TypeSystem::NaryOperator {
```
Comment (pmenon) [inline src/codegen/type/varchar_type.cpp]: Can you add some comments somewhere that outlines what the expected arguments to the function are?

File: test/function/functions_test.cpp [hunk 4]
```diff
@@ -97,22 +95,21 @@ TEST_F(FunctionsTests, FuncCallTest) {
...
-  EXPECT_EQ(result[1][0], 'a');
-  EXPECT_EQ(result[1][1], 'b');
+//  TestingSQLUtil::ExecuteSQLQuery("SELECT SQRT(a), SUBSTR(s,1,2) FROM test;",
+//                                  result, tuple_descriptor, rows_affected,
+//                                  error_message);
+//  EXPECT_EQ(1, result[0].size());
+//  EXPECT_EQ('2', result[0][0]);
+//  EXPECT_EQ(2, result[1].size());
+//  EXPECT_EQ(result[1][0], 'a');
+//  EXPECT_EQ(result[1][1], 'b');
```
Comment (pmenon) [inline test/function/functions_test.cpp]: Why is this commented out?

File: test/function/functions_test.cpp [hunk 5]
```diff
@@ -127,5 +124,52 @@ TEST_F(FunctionsTests, FuncCallTest) {
...
+  //  res = TestingSQLUtil::GetResultValueAsString(result, 0);
+  //  EXPECT_EQ("1", res);
+  //
+  //  TestingSQLUtil::ExecuteSQLQuery("SELECT SUBSTR(s,-2,2) FROM test;",
+  //  result,
+  //                                  tuple_descriptor, rows_affected,
+  //                                  error_message);
+  //  EXPECT_EQ(1, result.size());
+  //  res = TestingSQLUtil::GetResultValueAsString(result, 0);
+  //  EXPECT_EQ("", res);
```
Comment (pmenon) [inline test/function/functions_test.cpp]: Remove or undo commented code.

File: test/function/string_functions_test.cpp [hunk 6]
```diff
@@ -285,7 +285,7 @@ TEST_F(StringFunctionsTests, BTrimTest) {
     EXPECT_TRUE(result.IsNull());
   }
 }
-
+
```
Comment (pmenon) [inline test/function/string_functions_test.cpp]: Undo

File: test/function/functions_test.cpp [hunk 7]
```diff
@@ -9,16 +9,14 @@
 // Copyright (c) 2015-2017, Carnegie Mellon University Database Group
 //
 //===----------------------------------------------------------------------===//
-
```
Comment (pmenon) [inline test/function/functions_test.cpp]: Leave empty line after header.

File: src/codegen/type/varchar_type.cpp (hunk 3 above)
Comment (luochenUmich) [reply src/codegen/type/varchar_type.cpp]: Sure, just added some comments

Comment (luochenUmich) [reply src/codegen/type/varchar_type.cpp]: Just added some comments

File: test/function/functions_test.cpp (hunk 4 above)
Comment (luochenUmich) [reply test/function/functions_test.cpp]: Did it accidentally. Already uncommented it.

**Comment by luochenUmich** [comment]
@

**Comment by luochenUmich** [comment]
@pmenon All comments are addressed.

**Comment by pmenon** [comment]
@luochenUmich LGTM. Just run the formatter script on `varchar_type.cpp` and it should be good to go.

**Comment by luochenUmich** [comment]
@pmenon Done.

**Comment by luochenUmich** [comment]
@apavlo This should be good to go

**Comment by apavlo** [comment]
I will merge this next.

**Pull request merged**